src/network/networkd-netdev-wireguard.c

```diff
@@ -0,0 +1,625 @@
+        else
+                log_netdev_error(e->netdev, "Neither IPv4 nor IPv6 address found for peer endpoint: %s:%s", e->host, e->port);
+
+        w->n_unresolved_endpoints--;
+
+        if (!w->n_unresolved_endpoints)
+                update_wireguard_config(e->netdev);
+        else if (w->n_unresolved_endpoints % MAX_CONCURRENT_QUERIES == 0)
+                resolve_endpoints(e->netdev);
+
+        return 0;
+}
+
+static void resolve_endpoints(NetDev *netdev) {
+        int r, i = 0;
+        Wireguard *w;
+        WireguardEndpoint *endpoint;
+        struct addrinfo hints = {
+                .ai_socktype = SOCK_DGRAM,
+                .ai_protocol = IPPROTO_UDP,
+                .ai_flags = AI_ADDRCONFIG
+        };
+
+        assert(netdev);
+        w = WIREGUARD(netdev);
+        assert(w);
+
+        LIST_FOREACH(endpoints, endpoint, w->unresolved_endpoints) {
+                if (++i >= MAX_CONCURRENT_QUERIES)
+                        break;
+
+                endpoint->netdev = netdev;
+                hints.ai_family = endpoint->peer->address_family;
+                r = sd_resolve_getaddrinfo(netdev->manager->resolve,
+                                           &w->resolve_query,
+                                           endpoint->host,
+                                           endpoint->port,
+                                           &hints,
+                                           wireguard_resolve_handler,
+                                           endpoint);
+
+                if (r < 0)
+                        log_netdev_error_errno(netdev, -errno, "Failed create resolver: %m");
+        }
+}
+
+
+static int netdev_wireguard_post_create(NetDev *netdev, Link *link, sd_netlink_message *m) {
+        Wireguard *w;
+
+        assert(netdev);
+        w = WIREGUARD(netdev);
+        assert(w);
+
+        update_wireguard_config(netdev);
+        resolve_endpoints(netdev);
+        return 0;
+}
+
+int config_parse_wireguard_listen_port(const char *unit,
+                                       const char *filename,
+                                       unsigned line,
+                                       const char *section,
+                                       unsigned section_line,
+                                       const char *lvalue,
+                                       int ltype,
+                                       const char *rvalue,
+                                       void *data,
+                                       void *userdata) {
+        Wireguard *w;
+        uint16_t port_num = 0;
+        int r;
+
+        assert(rvalue);
+        assert(data);
+
+        // TODO replace with sd_resolve_getaddrinfo to support service names
+        r = safe_atou16(rvalue, &port_num);
+        if (r < 0)
+                log_syntax(unit, LOG_ERR, filename, line, r, "Invalid port specification, ignoring assignment: %s", rvalue);
```

Comment:
I am pretty sure in networkd we shouldn't accept `Port=0` if that has a special meaning. Instead Port= should accept an additional value "auto" or so (which needs proper documentation of course) which would then internally be passed as 0 to wireguard... Using "0" with special semantics is fine for low-level code, but I doubt we should expose this to the user.